File: apps/android/app/src/main/java/com/cursumi/app/core/api/ApiClient.kt
package com.cursumi.app.core.api

import com.cursumi.app.core.Config
import com.cursumi.app.core.auth.CookieJar
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlinx.serialization.KSerializer
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.serializer
import okhttp3.Call
import okhttp3.Callback
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

class ApiException(val status: Int, message: String?) : Exception(message ?: "HTTP $status")
class NetworkException(cause: Throwable) : Exception("No se pudo conectar. Revisa tu conexión.", cause)

/** JSON tolerante para toda la app. */
val AppJson = Json {
    ignoreUnknownKeys = true
    isLenient = true
    coerceInputValues = true
    explicitNulls = false
}

/**
 * Cliente HTTP hacia la API de Cursumi. Adjunta la cookie de sesión y las
 * cabecera que espera el plugin native-app del servidor. OkHttp no gestiona cookies:
 * así controlamos qué se guarda y la cookie nunca sale del dominio de Cursumi.
 */
class ApiClient(val jar: CookieJar, val baseUrl: String = Config.API_URL) {
    val http: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .followRedirects(false)
        .build()

    class ApiResponse(val status: Int, val body: String, val headers: okhttp3.Headers) {
        val isOk get() = status in 200..299
        inline fun <reified T> decode(): T = AppJson.decodeFromString(body)
        fun <T> decode(serializer: KSerializer<T>): T = AppJson.decodeFromString(serializer, body)

        /** Mensaje `{ error }` o `{ message }` del cuerpo, si lo hay. */
        val errorMessage: String?
            get() = runCatching {
                val obj = AppJson.parseToJsonElement(body) as? JsonObject
                (obj?.get("error") as? JsonPrimitive)?.content ?: (obj?.get("message") as? JsonPrimitive)?.content
            }.getOrNull()

        fun throwIfError(): ApiResponse {
            if (!isOk) throw ApiException(status, errorMessage)
            return this
        }
    }

    fun url(path: String, query: Map<String, String> = emptyMap()): HttpUrl {
        val b = "$baseUrl/${path.trimStart('/')}".toHttpUrl().newBuilder()
        query.forEach { (k, v) -> b.addQueryParameter(k, v) }
        return b.build()
    }

    suspend fun request(
        method: String,
        path: String,
        query: Map<String, String> = emptyMap(),
        jsonBody: JsonElement? = null,
        rawBody: RequestBody? = null,
        auth: Boolean = true,
    ): ApiResponse = withContext(Dispatchers.IO) {
        val builder = Request.Builder().url(url(path, query)).header("Accept", "application/json")
        if (auth) jar.cookieHeader()?.let { builder.header("Cookie", it) }
        if (path.trimStart('/').startsWith("api/auth/")) {
            // El plugin native-app del servidor copia esta cabecera a `Origin`.
            builder.header("x-native-origin", Config.ORIGIN)
        }
        val body: RequestBody? = when {
            jsonBody != null -> jsonBody.toString().toRequestBody("application/json".toMediaType())
            rawBody != null -> rawBody
            method == "POST" || method == "PATCH" || method == "PUT" -> "".toRequestBody(null)
            else -> null
        }
        builder.method(method, body)
        val response = try {
            http.newCall(builder.build()).await()
        } catch (e: IOException) {
            throw NetworkException(e)
        }
        response.use {
            jar.store(it.headers("Set-Cookie"), it.request.url)
            ApiResponse(it.code, it.body.string(), it.headers)
        }
    }

    suspend inline fun <reified T> get(path: String, query: Map<String, String> = emptyMap(), auth: Boolean = true): T =
        request("GET", path, query, auth = auth).throwIfError().decode()

    suspend inline fun <reified B, reified T> post(path: String, body: B): T =
        request("POST", path, jsonBody = AppJson.encodeToJsonElement(serializer<B>(), body)).throwIfError().decode()

    suspend inline fun <reified B> postUnit(path: String, body: B? = null) {
        request("POST", path, jsonBody = body?.let { AppJson.encodeToJsonElement(serializer<B>(), it) }).throwIfError()
    }

    suspend fun postEmpty(path: String) { request("POST", path).throwIfError() }

    suspend inline fun <reified B> patch(path: String, body: B? = null) {
        request("PATCH", path, jsonBody = body?.let { AppJson.encodeToJsonElement(serializer<B>(), it) }).throwIfError()
    }

    suspend inline fun <reified B> delete(path: String, body: B? = null) {
        request("DELETE", path, jsonBody = body?.let { AppJson.encodeToJsonElement(serializer<B>(), it) }).throwIfError()
    }
}

suspend fun Call.await(): Response = suspendCancellableCoroutine { cont ->
    enqueue(object : Callback {
        override fun onFailure(call: Call, e: IOException) { if (cont.isActive) cont.resumeWithException(e) }
        override fun onResponse(call: Call, response: Response) { cont.resume(response) }
    })
    cont.invokeOnCancellation { cancel() }
}

File: apps/ios/Cursumi/Core/API/APIClient.swift
import Foundation

enum APIError: LocalizedError {
    case http(Int, message: String?)
    case decoding(Error)
    case network(Error)

    var errorDescription: String? {
        switch self {
        case .http(let status, let message):
            return message ?? "HTTP \(status)"
        case .decoding:
            return "La respuesta del servidor no tiene el formato esperado."
        case .network:
            return "No se pudo conectar. Revisa tu conexión."
        }
    }
}

/// Cliente HTTP hacia la API de Cursumi. Adjunta la cookie de sesión y las
/// cabeceras que espera el plugin expo del servidor.
final class APIClient {
    static let shared = APIClient()

    let baseURL: URL
    let jar: CookieJar
    private let session: URLSession

    init(baseURL: URL = Config.apiURL, jar: CookieJar = CookieJar()) {
        self.baseURL = baseURL
        self.jar = jar
        // Las cookies las gestiona `CookieJar`, no Foundation: así controlamos qué
        // se guarda y evitamos que la cookie viaje a otros dominios.
        let config = URLSessionConfiguration.default
        config.httpShouldSetCookies = false
        config.httpCookieAcceptPolicy = .never
        config.timeoutIntervalForRequest = 30
        session = URLSession(configuration: config)
    }

    // MARK: - Peticiones

    struct Response {
        let status: Int
        let data: Data
        let headers: [AnyHashable: Any]

        var isOK: Bool { (200..<300).contains(status) }

        func decode<T: Decodable>(_ type: T.Type) throws -> T {
            do {
                return try JSONDecoder.api.decode(type, from: data)
            } catch {
                throw APIError.decoding(error)
            }
        }

        /// Mensaje `{ error: "..." }` o `{ message: "..." }` del cuerpo, si lo hay.
        var errorMessage: String? {
            guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
            return (obj["error"] as? String) ?? (obj["message"] as? String)
        }
    }

    func request(
        _ method: String,
        _ path: String,
        query: [String: String] = [:],
        json body: Encodable? = nil,
        rawBody: Data? = nil,
        contentType: String? = nil,
        auth: Bool = true
    ) async throws -> Response {
        var components = URLComponents(url: baseURL.appendingPathComponent(path), resolvingAgainstBaseURL: false)!
        if !query.isEmpty {
            components.queryItems = query.map { URLQueryItem(name: $0.key, value: $0.value) }
        }
        var req = URLRequest(url: components.url!)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        if auth, let cookie = jar.cookieHeader {
            req.setValue(cookie, forHTTPHeaderField: "Cookie")
        }
        if path.hasPrefix("api/auth/") || path.hasPrefix("/api/auth/") {
            // Mismas cabeceras que manda @better-auth/expo.
            req.setValue(Config.origin, forHTTPHeaderField: "expo-origin")
            req.setValue("true", forHTTPHeaderField: "x-skip-oauth-proxy")
        }
        if let body {
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONEncoder().encode(AnyEncodable(body))
        } else if let rawBody {
            req.httpBody = rawBody
            if let contentType { req.setValue(contentType, forHTTPHeaderField: "Content-Type") }
        }

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await session.data(for: req)
        } catch {
            throw APIError.network(error)
        }
        let http = response as? HTTPURLResponse
        let headers = http?.allHeaderFields ?? [:]
        jar.store(responseHeaders: headers, for: req.url!)
        return Response(status: http?.statusCode ?? 0, data: data, headers: headers)
    }

    // MARK: - Atajos

    func get<T: Decodable>(_ path: String, query: [String: String] = [:], auth: Bool = true) async throws -> T {
        let res = try await request("GET", path, query: query, auth: auth)
        guard res.isOK else { throw APIError.http(res.status, message: res.errorMessage) }
        return try res.decode(T.self)
    }

    @discardableResult
    func post<T: Decodable>(_ path: String, _ body: Encodable? = nil) async throws -> T {
        let res = try await request("POST", path, json: body)
        guard res.isOK else { throw APIError.http(res.status, message: res.errorMessage) }
        return try res.decode(T.self)
    }

    func post(_ path: String, _ body: Encodable? = nil) async throws {
        let res = try await request("POST", path, json: body)
        guard res.isOK else { throw APIError.http(res.status, message: res.errorMessage) }
    }

    func patch(_ path: String, _ body: Encodable? = nil) async throws {
        let res = try await request("PATCH", path, json: body)
        guard res.isOK else { throw APIError.http(res.status, message: res.errorMessage) }
    }

    func delete(_ path: String, _ body: Encodable? = nil) async throws {
        let res = try await request("DELETE", path, json: body)
        guard res.isOK else { throw APIError.http(res.status, message: res.errorMessage) }
    }
}

/// Envuelve cualquier `Encodable` para poder pasarlo como existencial.
struct AnyEncodable: Encodable {
    private let encode: (Encoder) throws -> Void
    init(_ value: Encodable) { encode = value.encode }
    func encode(to encoder: Encoder) throws { try encode(encoder) }
}

extension JSONDecoder {
    /// Decoder tolerante con las fechas ISO de la API (con y sin fracción de segundo).
    static let api: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .custom { decoder in
            let raw = try decoder.singleValueContainer().decode(String.self)
            if let date = ISO8601DateFormatter.fractional.date(from: raw) ?? ISO8601DateFormatter.plain.date(from: raw) {
                return date
            }
            throw DecodingError.dataCorrupted(.init(codingPath: decoder.codingPath, debugDescription: "Fecha inválida: \(raw)"))
        }
        return decoder
    }()
}

extension ISO8601DateFormatter {
    static let fractional: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()
    static let plain = ISO8601DateFormatter()
}
